flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,423 @@
+
+      // deserialize field 1
+      if (!nullMask1(i)) {
+        deserializedKeyFields1(i) = serializer.deserialize(deserializedKeyFields1(i), firstSource)
+      }
+
+      // deserialize field 2
+      if (!nullMask2(i)) {
+        deserializedKeyFields2(i) = serializer.deserialize(deserializedKeyFields2(i), secondSource)
+      }
+
+      i += 1
+    }
+
+    // compare
+    i = 0
+    while (i < keyLen) {
+      val keyPos = keyPositions(i)
+      val comparator = comparators(i).asInstanceOf[TypeComparator[Any]]
+
+      val isNull1 = nullMask1(keyPos)
+      val isNull2 = nullMask2(keyPos)
+
+      var cmp = 0
+      // both values are null -> equality
+      if (isNull1 && isNull2) {
+        cmp = 0
+      }
+      // first value is null -> inequality
+      else if (isNull1) {
+        cmp = comparator.compare(null, deserializedKeyFields2(keyPos))
+      }
+      // second value is null -> inequality
+      else if (isNull2) {
+        cmp = comparator.compare(deserializedKeyFields1(keyPos), null)
+      }
+      // no null values
+      else {
+        cmp = comparator.compare(deserializedKeyFields1(keyPos), deserializedKeyFields2(keyPos))
+      }
+
+      if (cmp != 0) {
+        return cmp
+      }
+
+      i += 1
+    }
+    0
+  }
+
+  override def supportsNormalizedKey(): Boolean = numLeadingNormalizableKeys > 0
+
+  override def getNormalizeKeyLen: Int = normalizableKeyPrefixLen
+
+  override def isNormalizedKeyPrefixOnly(keyBytes: Int): Boolean =
+    numLeadingNormalizableKeys < keyPositions.length ||
+      normalizableKeyPrefixLen == Integer.MAX_VALUE ||
+      normalizableKeyPrefixLen > keyBytes
+
+  override def invertNormalizedKey(): Boolean = invertNormKey
+
+  override def supportsSerializationWithKeyNormalization(): Boolean = false
+
+  override def writeWithKeyNormalization(record: Row, target: DataOutputView): Unit =
+    throw new UnsupportedOperationException("Record serialization with leading normalized keys " +
+      "not supported.")
+
+  override def readWithKeyDenormalization(reuse: Row, source: DataInputView): Row =
+    throw new UnsupportedOperationException("Record deserialization with leading normalized keys " +
+      "not supported.")
+
+  override def duplicate(): TypeComparator[Row] = {
+    // copy comparator and serializer factories
+    val comparatorsCopy = comparators.map(_.duplicate().asInstanceOf[NullAwareComparator[_]])
+    val serializersCopy = serializers.map(_.duplicate())
+
+    new RowComparator(
+      keyPositions,
+      comparatorsCopy,
+      serializersCopy,
+      auxiliaryFields)
+  }
+
+  override def hash(value: Row): Int = {
+    var code: Int = 0
+    var i = 0
+    try {
+      while(i < keyPositions.length) {
+        code *= TupleComparatorBase.HASH_SALT(i & 0x1F)
+        val element = value.productElement(keyPositions(i)) // element can be null
+        code += comparators(i).asInstanceOf[TypeComparator[Any]].hash(element)
```

Comment:
I think you can you get rid of these casts if you define `comparators` as `Array[NullAwareComparator[Any]]` instead of `Array[NullAwareComparator[_]]`.